native_client/BUILD

```diff
@@ -45,32 +45,58 @@ genrule(
                 exclude = ["kenlm/*/*test.cc", "kenlm/*/*main.cc"]) +
            glob(["boost_locale/**/*.hpp"]),
-    hdrs = ["deepspeech.h"],
     # -Wno-sign-compare to silent a lot of warnings from tensorflow itself,
     # which makes it harder to see our own warnings
-    copts = ["-std=c++11", "-Wno-sign-compare"] + if_native_model([
+    copts = ["-std=c++11", "-Wno-sign-compare", "-fvisibility=hidden"] + if_native_model([
         "-DDS_MODEL_TIMESTEPS=$(DS_MODEL_TIMESTEPS)",
         "-DDS_NATIVE_MODEL=1",
     ]),
-    linkopts = select({
-        "//tensorflow:darwin": [
-            "-Wl,-install_name,@rpath/libdeepspeech.so"
-	],
-	"//conditions:default": []
-    }),
     deps = [
-        "//tensorflow/core:core",
+        "//tensorflow/core:core_cpu",
+        "//tensorflow/core:direct_session",
+        ### "//tensorflow/core:all_kernels",
+        ### => Trying to be more fine-grained
+        ### Obtained by trial/error process ...
+        ### CPU only build libdeepspeech.so from 63M to 36M
+        "//tensorflow/core/kernels:constant_op",        # Const
+        "//tensorflow/core/kernels:identity_op",        # Identity
+        "//tensorflow/core/kernels:transpose_op",       # Transpose
+        "//tensorflow/core/kernels:reshape_op",         # Reshape
+        "//tensorflow/core/kernels:shape_ops",          # Shape
+        "//tensorflow/core/kernels:strided_slice_op",   # StridedSlice
+        "//tensorflow/core/kernels:pack_op",            # Pack
+        "//tensorflow/core/kernels:reverse_op",         # ReverseV2
+        "//tensorflow/core/kernels:concat_op",          # ConcatV2
+        "//tensorflow/core/kernels:split_op",           # Split
+        "//tensorflow/core/kernels:sparse_to_dense_op", # SparseToDense
+        "//tensorflow/core/kernels:relu_op",            # Relu
+        "//tensorflow/core/kernels:bias_op",            # BiasAdd
+        "//tensorflow/core/kernels:math",               # Range, MatMul
+        "//tensorflow/core/kernels:tensor_array_ops",   # Placeholder, TensorArrayV3
+        "//tensorflow/core/kernels:control_flow_ops",   # Enter
+        "//tensorflow/core/kernels:ctc_ops",            # CTCBeamSearchDecoder
+        ### Needed by production model produced without "--use_seq_length False"
+        "//tensorflow/core/kernels:logging_ops",         # Assert
+        "//tensorflow/core/kernels:reverse_sequence_op", # ReverseSequence
+        # Classic deps
         "//tensorflow/core/util/ctc",
         "//third_party/eigen3",
-        ":deepspeech_utils"
-    ] + if_native_model([":deepspeech_model", "//tensorflow/compiler/tf2xla:xla_compiled_cpu_function"]),
+    ] + if_native_model([
+        "//tensorflow/compiler/tf2xla:xla_compiled_cpu_function",
+        "//tensorflow/compiler/xla/service/cpu:runtime_matmul",
+    ])
+      + if_cuda([
+        "//tensorflow/core:core",
+        "//tensorflow/core/kernels:slice_op_gpu",       # Slice GPU
```

Comment:
Does `core:core` above include the GPU kernels for all the other OPs needed in the CPU list?

Reply:
`core:core` includes the GPU runtime. I was unable to run on GPU without `slice_op_gpu`, it complained of a missing kernel. Taking a look at higher level of debug with GPU would show a lot of ops executed on the GPU, but if you can check that as well on your side, it's welcome :).

My process was running with `TF_CPP_MIN_LOG_LEVEL` and then using `grep` / `wc` to cound number of occurrences with `GPU` and `CPU` with both runtimes, and compare to the old version. I'll do that again to make sure next monday.

Reply:
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech/gpu_onelib$ TF_CPP_MIN_VLOG_LEVEL=1 ./deepspeech /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/test.frozen.988_e120.LSTM.ldc93s1.pb /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/ldc93s1/LDC93S1_long.wav /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/alphabet.txt /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/lm.binary /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/trie -t 2>tf_vlog.txt
she had your dad your dad your dad your dad your dad your dark suit in greasy wash water all year
cpu_time_overall=1.50706 cpu_time_mfcc=0.00629 cpu_time_infer=1.50077
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech/gpu_onelib$ cd ../gpu_legacy/
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech/gpu_legacy$ TF_CPP_MIN_VLOG_LEVEL=1 ./deepspeech /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/test.frozen.988_e120.LSTM.ldc93s1.pb /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/ldc93s1/LDC93S1_long.wav /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/alphabet.txt /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/lm.binary /home/alexandre/Documents/codaz/Mozilla/DeepSpeech/DeepSpeech-lissyx/data/lm/trie -t 2>tf_vlog.txt
she had your dad your dad your dad your dad your dad your dark suit in greasy wash water all year
cpu_time_overall=1.65676 cpu_time_mfcc=0.00650 cpu_time_infer=1.65027

Reply:
And then, comparing runs:
```
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:GPU:0" gpu_legacy/tf_vlog.txt 
18077
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:CPU:0" gpu_legacy/tf_vlog.txt 
16
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ ls -hal gpu_legacy/libdeepspeech.so
-r-xr-xr-x 1 alexandre alexandre 788K janv.  5 11:50 gpu_legacy/libdeepspeech.so
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:GPU:0" gpu_onelib/tf_vlog.txt 
18077
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ grep -ic "device:CPU:0" gpu_onelib/tf_vlog.txt 
16
(tf-venv-master) alexandre@serveur:~/tmp/deepspeech$ ls -hal gpu_onelib/libdeepspeech.so
-r-xr-xr-x 1 alexandre alexandre 315M janv.  4 15:09 gpu_onelib/libdeepspeech.so
```